### DISCUSSION TITLE
why does all my output say frozen importlib._bootstrap?

### DISCUSSION THREAD

**Discussion by fohria** [pythonspeed/filprofiler#297] [Q&A] [answered]
hey! thanks for this profiler, it looks very useful, if i can figure out how to use it :)

i have a short script that, in short, generates a bunch of data, and then plots it. depending on how much i generate, memory use can be many gigabytes. so i'd like to profile it so i can find out when and where i may have some dataframes hanging around from function calls that i can delete when they're not needed anymore, like after i have dumped it to a file.

however, running it with fil, i get this:

![image](https://user-images.githubusercontent.com/8695061/151163298-e3e11e1e-b137-4240-bcb7-75de88761341.png)

the light pink on the left are the plotting calls, but what does it mean that it says `<frozen importlib._bootstrap>` all over?

tldr version of my code is:

```
data = generate_data(how_much)  # returns a pd.dataframe
figure = plotting_call(data)
```

(i've installed fil to the same conda env i use for the script, if that matters)

**Comment by itamarst** [comment]
1. The redder and wider the frame, the more memory.
2. You read the code as you would a traceback, so the right stack is a traceback that looks like this:

```
app_cmd.py:2 (<module>)
import numpy as np
... lots of stuff with importlib ...
```

In other words, it's the memory from importing NumPy.

Is there a third stack if you scroll to the right, not shown in the screenshot? If not, something strange (a bug?) is happening, because importing NumPy is not 2GB worth of RAM, so where is the rest of the memory?

**Reply by itamarst** [reply]
Could you attach the SVG file to this discussion?

**Comment by fohria** [comment]
Oh you're right there are more columns to the right, very thin ones. I removed the plotting stuff to make it easier to see (in the attached svg), but as you say it still looks like numpy for whatever reason uses the 2GB. In my mind it *should* be the `data` variable holding a pandas dataframe taking up the most memory.

Oh I forgot to mention, I'm running fil in Ubuntu through WSL2, maybe that causes some issue?

![peak-memory](https://user-images.githubusercontent.com/8695061/151377824-61ebac76-cf70-471f-bbf0-0a4d003dbec4.svg)

**Reply by itamarst** [reply]
Could you attach the code + data needed to reproduce this? There definitely seems to be a problem, but it's hard to see what it is.

If not, could you try running inside Docker, rather than WSL, and see if that makes a difference?

**Reply by itamarst** [reply]
Also, what version of Fil are you using? And how did you install it, pip or Conda?

**Comment by fohria** [comment]
So I went through this a bit more systematically with new conda environments etc. Also tested on macOS.

All below results are using the example code from the documentation: https://pythonspeed.com/fil/docs/fil/trying.html

Not sure if I'm doing something wrong but results are definitely not as easy to interpret as in the documentation example :)

#### WSL on Win10/Ubuntu20.04
```
mamba 0.18.2
conda 4.10.3
```

Created new conda environment, installed fil-profiler and numpy:
```
mamba create -n filtest
conda activate filtest
mamba install filprofiler numpy  # mamba uses conda-forge by default
```

That gets fil version `filprofiler               2022.01.1`

Ran the example: `fil-profile run fil-example.py`, with the result in this attached svg:
![peak-memory](https://user-images.githubusercontent.com/8695061/152335875-b2827faf-163e-4b4d-8ad7-d035594ac679.svg)

#### macOS 11.6.3
```
mamba 0.20.0
conda 4.11.0
```

Installed fil into new conda env like above, gets me the same `2022.01.1` version of fil and run the same example and get this result:

![peak-memory](https://user-images.githubusercontent.com/8695061/152337407-1aa71f8a-8705-4bd0-b1ed-86f8f4b8b8c2.svg)

Just to make sure it's not mamba doing things in slightly different ways than conda, i also created new env and installed with just conda commands:

```
conda create -n filtester
conda activate filtester
conda install -c conda-forge filprofiler numpy
```

But that gave exact same svg result (double checked path to `fil-profile` was correct with `which fil-profile`)

**Reply by itamarst** [reply]
Thank you for the extensive testing. Some sort of bug here, looks like; I will next do some testing on my own; my first hypothesis is that this is about Conda, so one thing you can try to get yourself personally unstuck is to try `pip install filprofiler` instead of Conda. I will try to fix this regardless, but that might get you going.

**Reply by itamarst** [reply]
Eyeballing those graphs, the numbers for allocations are always fairly consistent, so the NumPy import is what varies a lot. And testing on Linux, I get very different memory usage for importing NumPy via PyPI install and Conda-Forge install, but the regular code usage was consistent.

So as a first pass, it seems like if you ignore memory usage from imports you will plausibly get correct information. So next question is, what's up with imports.

**Comment by itamarst** [comment, marked as answer]
I've added a numpy-specific workaround to the latest release (2022.3.0). More broadly, tracking how much of an mmap() is actually used (since by default no RAM is used until you start writing to it, allowing sparse usage) is tracking by https://github.com/pythonspeed/filprofiler/issues/308.
